File: 0002_Add_Two_Numbers/solution.c
#include <stdlib.h>


typedef struct ListNode {
    int val;
    struct ListNode *next;
} ListNode;


ListNode * add_two_numbers(ListNode *l1, ListNode *l2) {
    ListNode *head_sum = NULL;
    ListNode *tail_sum = NULL;

    ListNode *cur_l1 = l1;
    ListNode *cur_l2 = l2;
    int carry = 0;

    while (cur_l1 || cur_l2 || carry) {
        int sum = carry;
        if (cur_l1) {
            sum += cur_l1->val;
            cur_l1 = cur_l1->next;
        }
        if (cur_l2) {
            sum += cur_l2->val;
            cur_l2 = cur_l2->next;
        }

        ListNode *new_node = (ListNode *)malloc(sizeof(ListNode));
        new_node->val = sum % 10;
        carry = sum / 10;

        if (!head_sum) {
            head_sum = tail_sum = new_node;
        } else {
            tail_sum->next = new_node;
            tail_sum = new_node;
        }
    }
    tail_sum->next = NULL;
    return head_sum;
}

File: 0002_Add_Two_Numbers/solution.ts
class ListNode {
    val: number;
    next: ListNode | null;

    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined ? 0 : val);
        this.next = (next === undefined ? null : next);
    }
}

function createListList(arr: number[]): ListNode | null {
    if (arr.length === 0) return null;

    let head = new ListNode(arr[0]);
    let current = head;
    for (let i = 1; i < arr.length; i++) {
        current.next = new ListNode(arr[i]);
        current = current.next;
    }
    return head;
}

function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
    let head = new ListNode(0);
    let current = head;
    let carry = 0;

    while (l1 != null || l2 != null || carry > 0) {
        let sum = carry;
        if (l1 !== null) {
            sum += l1.val;
            l1 = l1.next;
        }
        if (l2 !== null) {
            sum += l2.val;
            l2 = l2.next;
        }

        carry = Math.floor(sum / 10);
        current.next = new ListNode(sum % 10);
        current = current.next;
    }
    return head.next;
}